### PR TITLE
Use scpPrettify() for envelopes

### LINKED ISSUE #1092: Use hex string formatting for `SCPEnvelope` instead of `ubyte[]`
Currently a log statement will produce this output, which is... A bit much:

```
2020-08-04T01:10:45.9215890Z 2020-08-04 01:10:42,964 [32mInfo[0m [agora.consensus.protocol.Nominator] - Rejected invalid envelope: { statement: { nodeID: 0x4288806450db5562857d893f2839aa1ccf89dc4fa4f6ffd6adca29f7987eadd3, slotIndex: cast(const(__c_ulonglong)) 12, pledges: {"prepare":{"prepared":{"counter":3,"value":{"base":"[ 8, 0, 1, 239, 250, 11, 168, 157, 127, 35, 16, 237, 113, 124, 125, 49, 28, 80, 108, 75, 172, 20, 123, 201, 250, 222, 189, 183, 4, 186, 237, 19, 40, 109, 187, 25, 101, 10, 3, 25, 158, 223, 164, 175, 134, 111, 158, 60, 122, 155, 206, 187, 240, 104, 241, 174, 229, 115, 228, 207, 124, 254, 141, 199, 203, 159, 225, 0, 177, 27, 184, 166, 149, 119, 37, 154, 27, 3, 29, 195, 24, 99, 60, 143, 214, 131, 10, 27, 162, 134, 22, 110, 104, 135, 232, 246, 64, 75, 206, 66, 31, 138, 244, 12, 13, 67, 37, 156, 11, 61, 168, 196, 30, 176, 106, 120, 186, 207, 185, 140, 180, 162, 105, 135, 114, 195, 92, 232, 124, 29, 108, 14, 1, 255, 0, 4, 42, 86, 89, 69, 0, 0, 157, 2, 56, 224, 161, 113, 64, 11, 198, 214, 138, 157, 155, 49, 106, 205, 81, 9, 100, 145, 19, 160, 92, 40, 79, 66, 150, 210, 179, 1, 34, 245, 0, 1, 35, 113, 101, 180, 155, 148, 79, 224, 133, 40, 233, 178, 239, 226, 76, 251, 109, 99, 142, 237, 112, 177, 241, 134, 68, 92, 138, 198, 67, 207, 187, 155, 164, 148, 45, 136, 138, 168, 69, 24, 197, 203, 78, 103, 140, 211, 52, 189, 174, 95, 108, 227, 216, 37, 90, 64, 171, 112, 62, 224, 11, 206, 144, 12, 0, 247, 56, 140, 243, 190, 211, 188, 175, 238, 164, 184, 97, 67, 52, 89, 203, 211, 76, 236, 211, 96, 35, 96, 5, 153, 83, 65, 180, 153, 179, 19, 66, 28, 78, 225, 50, 1, 58, 26, 215, 254, 219, 148, 173, 48, 86, 220, 240, 212, 71, 230, 242, 230, 200, 29, 255, 71, 241, 58, 66, 67, 9, 92, 3, 1, 255, 0, 4, 42, 86, 89, 69, 0, 0, 157, 2, 56, 224, 161, 113, 64, 11, 198, 214, 138, 157, 155, 49, 106, 205, 81, 9, 100, 145, 19, 160, 92, 40, 79, 66, 150, 210, 179, 1, 34, 245, 0, 1, 188, 79, 202, 11, 164, 98, 90, 149, 40, 183, 165, 94, 124, 140, 41, 4, 72, 88, 48, 21, 245, 149, 150, 163, 6, 9, 232, 188, 126, 33, 99, 186, 110, 144, 86, 34, 91, 25, 83, 31, 210, 124, 196, 95, 188, 118, 12, 27, 131, 116, 73, 230, 244, 29, 45, 138, 57, 219, 2, 11, 224, 250, 196, 219, 0, 123, 192, 227, 199, 145, 75, 41, 107, 36, 39, 162, 251, 27, 224, 62, 4, 149, 59, 139, 166, 126, 10, 187, 18, 53, 39, 190, 104, 105, 40, 0, 166, 107, 27, 182, 96, 159, 207, 113, 212, 136, 111, 194, 58, 89, 47, 36, 206, 225, 197, 215, 57, 215, 198, 156, 36, 209, 163, 146, 108, 205, 117, 104, 9, 1, 255, 0, 4, 42, 86, 89, 69, 0, 0, 157, 2, 56, 224, 161, 113, 64, 11, 198, 214, 138, 157, 155, 49, 106, 205, 81, 9, 100, 145, 19, 160, 92, 40, 79, 66, 150, 210, 179, 1, 34, 245, 0, 1, 50, 230, 190, 103, 73, 87, 241, 116, 236, 243, 69, 65, 87, 143, 92, 161, 185, 97, 73, 250, 174, 185, 193, 229, 176, 221, 50, 13, 108, 127, 155, 221, 176, 7, 184, 63, 113, 195, 27, 136, 122, 15, 86, 251, 69, 154, 162, 250, 182, 66, 82, 76, 141, 197, 147, 157, 120, 163, 234, 195, 182, 150, 105, 56, 0, 201, 84, 221, 131, 193, 17, 175, 253, 81, 83, 130, 243, 87, 111, 164, 42, 207, 134, 252, 9, 14, 234, 227, 163, 250, 11, 115, 237, 143, 68, 162, 118, 182, 161, 153, 199, 126, 8, 70, 80, 12, 115, 141, 113, 191, 160, 50, 18, 34, 49, 192, 207, 11, 61, 111, 127, 192, 255, 196, 33, 182, 14, 210, 13, 1, 255, 0, 4, 42, 86, 89, 69, 0, 0, 157, 2, 56, 224, 161, 113, 64, 11, 198, 214, 138, 157, 155, 49, 106, 205, 81, 9, 100, 145, 19, 160, 92, 40, 79, 66, 150, 210, 179, 1, 34, 245, 0, 1, 212, 203, 55, 112, 86, 174, 202, 41, 130, 73, 111, 188, 153, 188, 158, 91, 150, 28, 179, 69, 92, 152, 11, 156, 209, 184, 94, 9, 161, 54, 233, 165, 115, 140, 21, 238, 173, 114, 123, 72, 226, 113, 186, 17, 210, 78, 125, 118, 88, 219, 242, 127, 150, 217, 214, 10, 110, 80, 187, 161, 4, 39, 214, 172, 0, 74, 34, 233, 204, 210, 174, 97, 55, 109, 251, 99, 180, 205, 248, 210, 14, 84, 238, 198, 236, 122, 84, 252, 234, 197, 228, 113, 111, 250, 9, 204, 254, 60, 116, 194, 22, 141, 84, 38, 1, 38, 141, 88, 76, 101, 29, 90, 15, 69, 126, 109, 115, 61, 155, 199, 205, 108, 1, 31, 150, 134, 12, 133, 15, 1, 255, 0, 4, 42, 86, 89, 69, 0, 0, 157, 2, 56, 224, 161, 113, 64, 11, 198, 214, 138, 157, 155, 49, 106, 205, 81, 9, 100, 145, 19, 160, 92, 40, 79, 66, 150, 210, 179, 1, 34, 245, 0, 1, 137, 71, 104, 90, 46, 227, 130, 0, 16, 249, 133, 173, 109, 252, 89, 130, 18, 17, 90, 140, 205, 29, 229, 180, 113, 250, 44, 86, 94, 179, 125, 240, 131, 167, 142, 196, 174, 85, 187, 225, 224, 212, 239, 5, 156, 102, 0, 116, 209, 231, 159, 88, 19, 23, 38, 96, 214, 15, 222, 132, 75, 44, 34, 134, 0, 83, 8, 29, 202, 64, 15, 15, 168, 127, 76, 56, 62, 69, 26, 120, 43, 45, 137, 81, 140, 78, 144, 54, 27, 151, 112, 151, 123, 95, 200, 84, 140, 74, 38, 31, 228, 193, 68, 81, 74, 80, 116, 139, 161, 223, 113, 191, 45, 114, 134, 57, 76, 13, 34, 58, 5, 22, 50, 76, 197, 237, 78, 19, 7, 1, 255, 0, 4, 42, 86, 89, 69, 0, 0, 157, 2, 56, 224, 161, 113, 64, 11, 198, 214, 138, 157, 155, 49, 106, 205, 81, 9, 100, 145, 19, 160, 92, 40, 79, 66, 150, 210, 179, 1, 34, 245, 0, 1, 121, 28, 171, 166, 102, 47, 178, 67, 99, 98, 99, 79, 130, 23, 213, 44, 189, 237, 130, 236, 50, 80, 85, 109, 1, 0, 161, 247, 119, 149, 51, 59, 149, 6, 146, 148, 151, 226, 252, 70, 56, 52, 112, 35, 253, 197, 76, 86, 39, 157, 228, 240, 219, 231, 37, 250, 53, 232, 57, 129, 17, 194, 115, 201, 0, 113, 215, 85, 49, 107, 12, 73, 71, 164, 207, 253, 117, 241, 136, 225, 216, 185, 63, 190, 68, 239, 2, 141, 208, 166, 182, 170, 207, 229, 8, 181, 235, 34, 94, 22, 246, 151, 75, 17, 61, 140, 212, 30, 220, 196, 39, 52, 150, 127, 140, 131, 154, 193, 183, 43, 28, 158, 37, 66, 210, 229, 69, 167, 1, 1, 255, 0, 4, 42, 86, 89, 69, 0, 0, 157, 2, 56, 224, 161, 113, 64, 11, 198, 214, 138, 157, 155, 49, 106, 205, 81, 9, 100, 145, 19, 160, 92, 40, 79, 66, 150, 210, 179, 1, 34, 245, 0, 1, 1, 100, 221, 66, 225, 38, 37, 103, 56, 223, 207, 161, 65, 62, 30, 251, 42, 107, 165, 153, 6, 95, 182, 215, 214, 243, 200, 6, 128, 227, 42, 88, 200, 249, 16, 168, 26, 33, 224, 249, 26, 123, 222, 88, 89, 123, 59, 62, 255, 106, 166, 61, 107, 101, 198, 253, 163, 25, 125, 191, 19, 36, 135, 137, 0, 200, 144, 145, 251, 0, 174, 238, 133, 163, 82, 123, 55, 51, 174, 74, 41, 49, 174, 235, 113, 216, 59, 214, 33, 223, 58, 252, 215, 142, 74, 36, 152, 36, 206, 243, 169, 67, 251, 39, 154, 242, 56, 22, 218, 99, 26, 187, 132, 71, 244, 95, 5, 206, 178, 72, 181, 180, 146, 168, 126, 18, 86, 238, 10, 1, 255, 0, 4, 42, 86, 89, 69, 0, 0, 157, 2, 56, 224, 161, 113, 64, 11, 198, 214, 138, 157, 155, 49, 106, 205, 81, 9, 100, 145, 19, 160, 92, 40, 79, 66, 150, 210, 179, 1, 34, 245, 0 ]"}},"nC":0,"nH":3,"quorumSetHash":{"base":"0x8bdb3e59a9a76dd87b975ca4875f05a012b57eeb02e1a1d2e0d1fa5e99504143b7acff406efd7d12bbecde944f51b3e854029a8321ba51bc7b0b32bad4815261"},"preparedPrime":null,"ballot":{"counter":4,"value":{"base":"[ 8, 0, 1, 239, 250, 11, 168, 157, 127, 35, 16, 237, 113, 124, 125, 49, 28, 80, 108, 75, 172, 20, 123, 201, 250, 222, 189, 183, 4, 186, 237, 19, 40, 109, 187, 25, 101, 10, 3, 25, 158, 223, 164, 175, 134, 111, 158, 60, 122, 155, 206, 187, 240, 104, 241, 174, 229, 115, 228, 207, 124, 254, 141, 199, 203, 159, 225, 0, 177, 27, 184, 166, 149, 119, 37, 154, 27, 3, 29, 195, 24, 99, 60, 143, 214, 131, 10, 27, 162, 134, 22, 110, 104, 135, 232, 246, 64, 75, 206, 66, 31, 138, 244, 12, 13, 67, 37, 156, 11, 61, 168, 196, 30, 176, 106, 120, 186, 207, 185, 140, 180, 162, 105, 135, 114, 195, 92, 232, 124, 29, 108, 14, 1, 255, 0, 4, 42, 86, 89, 69, 0, 0, 157, 2, 56, 224, 161, 113, 64, 11, 198, 214, 138, 157, 155, 49, 106, 205, 81, 9, 100, 145, 19, 160, 92, 40, 79, 66, 150, 210, 179, 1, 34, 245, 0, 1, 35, 113, 101, 180, 155, 148, 79, 224, 133, 40, 233, 178, 239, 226, 76, 251, 109, 99, 142, 237, 112, 177, 241, 134, 68, 92, 138, 198, 67, 207, 187, 155, 164, 148, 45, 136, 138, 168, 69, 24, 197, 203, 78, 103, 140, 211, 52, 189, 174, 95, 108, 227, 216, 37, 90, 64, 171, 112, 62, 224, 11, 206, 144, 12, 0, 247, 56, 140, 243, 190, 211, 188, 175, 238, 164, 184, 97, 67, 52, 89, 203, 211, 76, 236, 211, 96, 35, 96, 5, 153, 83, 65, 180, 153, 179, 19, 66, 28, 78, 225, 50, 1, 58, 26, 215, 254, 219, 148, 173, 48, 86, 220, 240, 212, 71, 230, 242, 230, 200, 29, 255, 71, 241, 58, 66, 67, 9, 92, 3, 1, 255, 0, 4, 42, 86, 89, 69, 0, 0, 157, 2, 56, 224, 161, 113, 64, 11, 198, 214, 138, 157, 155, 49, 106, 205, 81, 9, 100, 145, 19, 160, 92, 40, 79, 66, 150, 210, 179, 1, 34, 245, 0, 1, 188, 79, 202, 11, 164, 98, 90, 149, 40, 183, 165, 94, 124, 140, 41, 4, 72, 88, 48, 21, 245, 149, 150, 163, 6, 9, 232, 188, 126, 33, 99, 186, 110, 144, 86, 34, 91, 25, 83, 31, 210, 124, 196, 95, 188, 118, 12, 27, 131, 116, 73, 230, 244, 29, 45, 138, 57, 219, 2, 11, 224, 250, 196, 219, 0, 123, 192, 227, 199, 145, 75, 41, 107, 36, 39, 162, 251, 27, 224, 62, 4, 149, 59, 139, 166, 126, 10, 187, 18, 53, 39, 190, 104, 105, 40, 0, 166, 107, 27, 182, 96, 159, 207, 113, 212, 136, 111, 194, 58, 89, 47, 36, 206, 225, 197, 215, 57, 215, 198, 156, 36, 209, 163, 146, 108, 205, 117, 104, 9, 1, 255, 0, 4, 42, 86, 89, 69, 0, 0, 157, 2, 56, 224, 161, 113, 64, 11, 198, 214, 138, 157, 155, 49, 106, 205, 81, 9, 100, 145, 19, 160, 92, 40, 79, 66, 150, 210, 179, 1, 34, 245, 0, 1, 50, 230, 190, 103, 73, 87, 241, 116, 236, 243, 69, 65, 87, 143, 92, 161, 185, 97, 73, 250, 174, 185, 193, 229, 176, 221, 50, 13, 108, 127, 155, 221, 176, 7, 184, 63, 113, 195, 27, 136, 122, 15, 86, 251, 69, 154, 162, 250, 182, 66, 82, 76, 141, 197, 147, 157, 120, 163, 234, 195, 182, 150, 105, 56, 0, 201, 84, 221, 131, 193, 17, 175, 253, 81, 83, 130, 243, 87, 111, 164, 42, 207, 134, 252, 9, 14, 234, 227, 163, 250, 11, 115, 237, 143, 68, 162, 118, 182, 161, 153, 199, 126, 8, 70, 80, 12, 115, 141, 113, 191, 160, 50, 18, 34, 49, 192, 207, 11, 61, 111, 127, 192, 255, 196, 33, 182, 14, 210, 13, 1, 255, 0, 4, 42, 86, 89, 69, 0, 0, 157, 2, 56, 224, 161, 113, 64, 11, 198, 214, 138, 157, 155, 49, 106, 205, 81, 9, 100, 145, 19, 160, 92, 40, 79, 66, 150, 210, 179, 1, 34, 245, 0, 1, 212, 203, 55, 112, 86, 174, 202, 41, 130, 73, 111, 188, 153, 188, 158, 91, 150, 28, 179, 69, 92, 152, 11, 156, 209, 184, 94, 9, 161, 54, 233, 165, 115, 140, 21, 238, 173, 114, 123, 72, 226, 113, 186, 17, 210, 78, 125, 118, 88, 219, 242, 127, 150, 217, 214, 10, 110, 80, 187, 161, 4, 39, 214, 172, 0, 74, 34, 233, 204, 210, 174, 97, 55, 109, 251, 99, 180, 205, 248, 210, 14, 84, 238, 198, 236, 122, 84, 252, 234, 197, 228, 113, 111, 250, 9, 204, 254, 60, 116, 194, 22, 141, 84, 38, 1, 38, 141, 88, 76, 101, 29, 90, 15, 69, 126, 109, 115, 61, 155, 199, 205, 108, 1, 31, 150, 134, 12, 133, 15, 1, 255, 0, 4, 42, 86, 89, 69, 0, 0, 157, 2, 56, 224, 161, 113, 64, 11, 198, 214, 138, 157, 155, 49, 106, 205, 81, 9, 100, 145, 19, 160, 92, 40, 79, 66, 150, 210, 179, 1, 34, 245, 0, 1, 137, 71, 104, 90, 46, 227, 130, 0, 16, 249, 133, 173, 109, 252, 89, 130, 18, 17, 90, 140, 205, 29, 229, 180, 113, 250, 44, 86, 94, 179, 125, 240, 131, 167, 142, 196, 174, 85, 187, 225, 224, 212, 239, 5, 156, 102, 0, 116, 209, 231, 159, 88, 19, 23, 38, 96, 214, 15, 222, 132, 75, 44, 34, 134, 0, 83, 8, 29, 202, 64, 15, 15, 168, 127, 76, 56, 62, 69, 26, 120, 43, 45, 137, 81, 140, 78, 144, 54, 27, 151, 112, 151, 123, 95, 200, 84, 140, 74, 38, 31, 228, 193, 68, 81, 74, 80, 116, 139, 161, 223, 113, 191, 45, 114, 134, 57, 76, 13, 34, 58, 5, 22, 50, 76, 197, 237, 78, 19, 7, 1, 255, 0, 4, 42, 86, 89, 69, 0, 0, 157, 2, 56, 224, 161, 113, 64, 11, 198, 214, 138, 157, 155, 49, 106, 205, 81, 9, 100, 145, 19, 160, 92, 40, 79, 66, 150, 210, 179, 1, 34, 245, 0, 1, 121, 28, 171, 166, 102, 47, 178, 67, 99, 98, 99, 79, 130, 23, 213, 44, 189, 237, 130, 236, 50, 80, 85, 109, 1, 0, 161, 247, 119, 149, 51, 59, 149, 6, 146, 148, 151, 226, 252, 70, 56, 52, 112, 35, 253, 197, 76, 86, 39, 157, 228, 240, 219, 231, 37, 250, 53, 232, 57, 129, 17, 194, 115, 201, 0, 113, 215, 85, 49, 107, 12, 73, 71, 164, 207, 253, 117, 241, 136, 225, 216, 185, 63, 190, 68, 239, 2, 141, 208, 166, 182, 170, 207, 229, 8, 181, 235, 34, 94, 22, 246, 151, 75, 17, 61, 140, 212, 30, 220, 196, 39, 52, 150, 127, 140, 131, 154, 193, 183, 43, 28, 158, 37, 66, 210, 229, 69, 167, 1, 1, 255, 0, 4, 42, 86, 89, 69, 0, 0, 157, 2, 56, 224, 161, 113, 64, 11, 198, 214, 138, 157, 155, 49, 106, 205, 81, 9, 100, 145, 19, 160, 92, 40, 79, 66, 150, 210, 179, 1, 34, 245, 0, 1, 1, 100, 221, 66, 225, 38, 37, 103, 56, 223, 207, 161, 65, 62, 30, 251, 42, 107, 165, 153, 6, 95, 182, 215, 214, 243, 200, 6, 128, 227, 42, 88, 200, 249, 16, 168, 26, 33, 224, 249, 26, 123, 222, 88, 89, 123, 59, 62, 255, 106, 166, 61, 107, 101, 198, 253, 163, 25, 125, 191, 19, 36, 135, 137, 0, 200, 144, 145, 251, 0, 174, 238, 133, 163, 82, 123, 55, 51, 174, 74, 41, 49, 174, 235, 113, 216, 59, 214, 33, 223, 58, 252, 215, 142, 74, 36, 152, 36, 206, 243, 169, 67, 251, 39, 154, 242, 56, 22, 218, 99, 26, 187, 132, 71, 244, 95, 5, 206, 178, 72, 181, 180, 146, 168, 126, 18, 86, 238, 10, 1, 255, 0, 4, 42, 86, 89, 69, 0, 0, 157, 2, 56, 224, 161, 113, 64, 11, 198, 214, 138, 157, 155, 49, 106, 205, 81, 9, 100, 145, 19, 160, 92, 40, 79, 66, 150, 210, 179, 1, 34, 245, 0 ]"}}}} }, signature: 0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 }
```

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
Fixes #1092

### CODE REVIEW

**Comment by AndrejMitrovic** [comment]
@bpfkorea/core green.

File: source/agora/consensus/protocol/Nominator.d [hunk 1]
```diff
@@ -302,12 +302,12 @@ extern(D):
...
-            log.trace("Rejected envelope with outdated slot #{} (ledger: #{})",
-                envelope.statement.slotIndex, cur_height);
-            return;  // slot was already externalized, ignore outdated message
-        }
+        //if (envelope.statement.slotIndex <= cur_height)
+        //{
+        //    log.trace("Rejected envelope with outdated slot #{} (ledger: #{})",
+        //        envelope.statement.slotIndex, cur_height);
+        //    return;  // slot was already externalized, ignore outdated message
+        //}
```
Comment (Geod24) [inline source/agora/consensus/protocol/Nominator.d]: ?

File: source/agora/consensus/protocol/Nominator.d [hunk 2]
```diff
@@ -27,7 +27,7 @@ import agora.consensus.data.Transaction;
 import agora.network.NetworkManager;
 import agora.node.Ledger;
 import agora.utils.Log;
-import agora.utils.PrettyPrinter;
+import agora.utils.SCPPrettyPrinter;
```
Comment (Geod24) [inline source/agora/consensus/protocol/Nominator.d:30]: Dead import ?

File: source/agora/consensus/protocol/Nominator.d (hunk 1 above)
Comment (AndrejMitrovic) [reply source/agora/consensus/protocol/Nominator.d]: Ah I was just testing it, I'll revert this.

**Comment by AndrejMitrovic** [comment]
Sorry I didn't push again after the rebase. Should be fixed now.

**Review by Geod24 — APPROVED** [review]

**Pull request merged**